components/brave_new_tab_ui/reducers/new_tab_reducer.tsx

```diff
@@ -7,28 +7,45 @@ import { Reducer } from 'redux'
 // Constants
 import { types } from '../constants/new_tab_types'
 import { Preferences } from '../api/preferences'
+import { Stats } from '../api/stats'
 
 // API
+import * as backgroundAPI from '../api/background'
 import * as gridAPI from '../api/topSites/grid'
-import * as dataFetchAPI from '../api/dataFetch'
+import { InitialData } from '../api/initialData'
 import * as bookmarksAPI from '../api/topSites/bookmarks'
 import * as dndAPI from '../api/topSites/dnd'
 import * as storage from '../storage'
 
+const initialState = storage.load()
+
 export const newTabReducer: Reducer<NewTab.State | undefined> = (state: NewTab.State | undefined, action: any) => {
+  console.timeStamp('reducer ' + action.type)
   if (state === undefined) {
-    state = storage.load()
-
-    setImmediate(() => {
-      dataFetchAPI.fetchTopSites()
-    })
+    console.timeStamp('reducer init')
+    state = initialState
   }
 
   const startingState = state
   const payload = action.payload
   switch (action.type) {
-    case types.NEW_TAB_TEXT_DIRECTION_UPDATED:
-      state = { ...state, textDirection: payload }
+    case types.NEW_TAB_SET_INITIAL_DATA:
+      const initialDataPayload = payload as InitialData
+      state = {
+        ...state,
+        initialDataLoaded: true,
+        ...initialDataPayload.preferences,
+        stats: initialDataPayload.stats,
+        ...initialDataPayload.privateTabData,
+        topSites: initialDataPayload.topSites
+      }
+      if (initialDataPayload.preferences.showBackgroundImage) {
+        state.backgroundImage = backgroundAPI.randomBackgroundImage()
+      }
+      console.timeStamp('reducer initial data received')
+      window.setTimeout(function () {
```

Comment:
I moved it from [the removed `case types.NEW_TAB_TOP_SITES_DATA_UPDATED:`](https://github.com/brave/brave-core/pull/2957/files/72684d0c7467d26388d35436f5d6abbd37f26cad#diff-d262f4f98d5c2050430c5e0dec4bdf8bL54), which is no longer required as we dispatch a single action for all initial data. I noticed in that removed case, the re-dispatch was not behind a `setImmediate`, meaning the action could be dispatched synchronously before this action has reduced to state and rendered.

I forgot that webpack does inject a `setImmediate` polyfill for web, if you feel we should use that.

Reply:
no need to polyfill but seems valid to add a comment referencing why it is needed. for future reference

Reply:
I ended up changing all the occurances of this anti-pattern of dispatching an action inside a reducer before the reducer has finished. It can introduce weird bugs where state changes go missing. See discussion at https://stackoverflow.com/questions/36730793/can-i-dispatch-an-action-in-reducer for more information.

I wrapped them in a `setImmediate` and left a TODO comment to refactor.